src/util/bip152.rs

```diff
@@ -79,6 +137,39 @@ pub struct HeaderAndShortIds {
 }
 impl_consensus_encoding!(HeaderAndShortIds, header, nonce, short_ids, prefilled_txs);
 
+impl HeaderAndShortIds {
+    /// Create a new HeaderAndShortIds from a full block.
+    ///
+    /// > Nodes SHOULD NOT use the same nonce across multiple different blocks.
+    #[cfg(feature = "siphasher")]
+    pub fn from_block(block: &Block, nonce: u64, use_wtxid: bool) -> HeaderAndShortIds {
+        use util::hash::BitcoinHash;
+
+        let siphash_keys = ShortId::calculate_siphash_keys(&block.header, nonce);
+
+        HeaderAndShortIds {
+            header: block.header.clone(),
+            nonce: nonce,
+            // Provide coinbase prefilled.
+            prefilled_txs: vec![PrefilledTransaction(0, block.txdata[0].clone())],
```

Comment:
This seems like hard coding a specific scenario for block transmission, but I'm not sure that's really a meaningful limitation in practice.

Perhaps a variant of `from_block` is warranted, with a parameter for specifying which transactions to prefill. I tried to do this based on a `Fn(usize, &Transaction) -> bool` argument, and also considered specifying which indexes to prefill, but I couldn't find a simple enough solution to make it seem justifiable.

I can think of two use cases where doing this could eliminate a block transactions request, both of which seem a bit contrived:

1. an edge node that relays txs and blocks to whitelisted, non-public nodes, and keeps track of which transactions each node knows about.
2. mining setting, where the miner knows some txs haven't been broadcast yet and wants to ensure their block can be validated without additional round trips

Reply:
Can you make a clear summary of what functionality you are missing?

I have to admit I built the interfacing functions based on some RPC functionality we use in Elements. (We added some stateless RPC calls to use the C++ CompactBlocks code, which we are calling from a Rust program, so I ported them to native Rust instead.)

Reply:
I'm changing `from_block` to a more general form taking a `&[bool]` to indicate which txs to prefill.